elasticdl/python/worker/task_data_service.py

```diff
@@ -27,44 +32,67 @@ def __init__(
+        self._cur_record_count = 0
         self._reported_record_count = 0
-        self._pending_tasks_with_counts = []
+        self._failed_record_count = 0
+        self._pending_tasks = deque()
         self._current_task = None
 
     def get_current_task(self):
         return self._current_task
 
+    def report_task_done(self, task_id, err_msg):
+        self._worker.report_task_result(task_id, err_msg)
+
     def report_record_done(self, count, err_msg=""):
         """
         Report the number of records in the latest processed batch,
         so TaskDataService knows if some pending tasks are finished
         and report_task_result to the master.
-        self._pending_tasks_with_counts[0][0] is the first pending task,
-        self._pending_tasks_with_counts[0][1] is the number of records
-        in this task.
         """
         self._reported_record_count += count
-        if (
-            len(self._pending_tasks_with_counts)
-            and self._reported_record_count
-            >= self._pending_tasks_with_counts[0][1]
-        ):
+        if err_msg:
+            self._failed_record_count += count
+            logger.warning(
+                "Process records of task contains " "error: %s", err_msg
+            )
+        if self._pending_tasks:
+            task = self._pending_tasks[0]
+            total_record_num = task.end - task.start
+            if task.type == elasticdl_pb2.TRAINING:
+                # only training, we allow for failures
+                if (
+                    self._failed_record_count / total_record_num
+                    > self._record_failure_tolerance
+                ):
+                    msg = (
+                        "too many records ({f}/{t}) failure, possible "
+                        "reason {err_msg}"
+                    ).format(
+                        err_msg=err_msg,
+                        f=self._failed_record_count,
+                        t=total_record_num,
+                    )
+                    self.report_task_done(task.task_id, msg)
+            elif self._failed_record_count != 0:
+                self.report_task_done(task.task_id, err_msg)
             with self._lock:
-                while (
-                    len(self._pending_tasks_with_counts)
-                    and self._reported_record_count
-                    >= self._pending_tasks_with_counts[0][1]
+                if (
+                    self._pending_tasks
+                    and self._reported_record_count >= total_record_num
                 ):
-                    task, _ = self._pending_tasks_with_counts.pop(0)
-                    self._worker.report_task_result(task.task_id, err_msg)
-                if len(self._pending_tasks_with_counts):
-                    self._current_task = self._pending_tasks_with_counts[0][0]
+                    task = self._pending_tasks.popleft()
+                    self.report_task_done(task.task_id, err_msg)
```

Comment:
Will we report twice for the same task in this method?

Reply:
Fixed